src/tasks/task_list.hpp

```diff
@@ -36,6 +36,75 @@ namespace parthenon {
 
 enum class TaskListStatus { running, stuck, complete, nothing_to_do };
 
+class TaskList;
+namespace task_list_impl {
+TaskID AddTaskHelper(TaskList *, Task);
+} // namespace task_list_impl
+
+class IterativeTasks {
+ public:
+  IterativeTasks(TaskList *tl, int key) : tl_(tl), key_(key) {
+    max_iterations_ = std::numeric_limits<unsigned int>::max();
+  }
+
+  template <class T, class... Args>
+  TaskID AddTask(TaskID const &dep, T &&func, Args &&... args) {
+    return AddTask_(TaskType::iterative, dep, std::forward<T>(func),
+                    std::forward<Args>(args)...);
+  }
+  // overload to add member functions of class T to task list
+  // NOTE: we must capture the object pointer
+  template <class T, class... Args>
+  TaskID AddTask(TaskID const &dep, TaskStatus (T::*func)(Args...), T *obj,
+                 Args &&... args) {
+    return this->AddTask_(TaskType::iterative, dep, [=]() mutable -> TaskStatus {
+      return (obj->*func)(std::forward<Args>(args)...);
+    });
+  }
+
+  template <class T, class... Args>
+  TaskID AddCompletionTask(TaskID const &dep, T &&func, Args &&... args) {
```

Comment:
Should this be `SetCompletionTask`? Can we ever have more than one completion task? How does that work?

Reply:
yeah that's fair.  I'll change it

Reply:
We might also wanna check that there's only ever one completion task.

Reply:
(Per set of iterative tasks, obviously)